### PR TITLE
Feature/764 state migrator

### PR DESCRIPTION
**Pull request by zzeekk**
see #764

### CODE REVIEW

File: sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRunState.scala [hunk 1]
```diff
@@ -125,11 +133,47 @@ private[smartdatalake] object ActionDAGRunState {
...
+      logger.info(s"Applying state migrators ${migrators.mkString(", ")} to state json for app=$appName runId=$runId attemptId=$attemptId")
+      Some(migrators.foldLeft(jValue)((v, m) => m.migrate(v)))
+    } else None
+  }
+
+  // list of state migrators, sorted in ascending order
+  private val stateMigrators: Seq[StateMigratorDef] = Seq(
+    new StateMigratorDef3To4()
+  ).sortBy(_.versionFrom) // force ordering
+  assert(stateMigrators.groupBy(_.versionFrom).forall(_._2.size == 1)) // check that versionFrom is unigue
```
Comment (pgruetter) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRunState.scala]:
```suggestion
  assert(stateMigrators.groupBy(_.versionFrom).forall(_._2.size == 1)) // check that versionFrom is unique
```

File: sdl-core/src/main/scala/io/smartdatalake/workflow/StateMigratorDef.scala [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  def versionTo: Int
+  def migrate(json: JValue): JValue
+  override def toString: String = s"$versionFrom -> $versionTo"
+}
+
+/**
+ * Migrate state from format version 3 to 4:
+ * - the structure of `actionState.results` was cleaned up.
+ *   instead of having `results.subFeed`, the attributes of `subFeed` have been moved directly into `result`.
+ * - Additionally `actionState.inputIds[].id` was cleaned up as `inputIds[]`, and the the same applies for `actionState.outputIds`.
```
Comment (pgruetter) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/StateMigratorDef.scala]:
```suggestion
 * - Additionally `actionState.inputIds[].id` was cleaned up as `inputIds[]`, and the same applies for `actionState.outputIds`.
```

File: sdl-core/src/main/scala/io/smartdatalake/workflow/StateMigratorDef.scala [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+ */
+class StateMigratorDef3To4 extends StateMigratorDef with SmartDataLakeLogger {
+  override val versionFrom = 3
+  override val versionTo = 4
+  override def migrate(json: JValue): JValue = {
+    assert(json \ "runStateFormatVersion" match { case JInt(version) => version <= versionFrom}, s"Version should be equals or less than $versionFrom")
```
Comment (pgruetter) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/StateMigratorDef.scala]: Shouldn't we assume that version = versionFrom ? The upgrade procedure is probably strictly working for one version only.

File: sdl-lang/src/main/scala/io/smartdatalake/meta/state/StateMigrator.scala [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+import io.smartdatalake.util.hdfs.HdfsUtil
+import io.smartdatalake.util.hdfs.HdfsUtil.RemoteIteratorWrapper
+import io.smartdatalake.util.misc.SmartDataLakeLogger
+import io.smartdatalake.workflow.ActionDAGRunState.checkStateFormatVersionAndMigrate
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs.{FileSystem, Path}
+import org.json4s.jackson.{JsonMethods, prettyJson}
+import scopt.OptionParser
+
+case class StateMigratorConfig(statePath: Option[String] = None)
```
Comment (pgruetter) [inline sdl-lang/src/main/scala/io/smartdatalake/meta/state/StateMigrator.scala:38]: Can you please add an overall ScalaDoc ?

**Review by pgruetter — COMMENTED** [review]
Great feature, just added a few comments. 
The assumption should be, that we strictly update one version to the next IMHO.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/StateMigratorDef.scala (hunk 3 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/StateMigratorDef.scala]: Then we could not migrate from version 1/2 to version 4. As there are no changes needed to migrate from version 1to2 or 2to3, no StateMigratorDef exists for these steps.

File: sdl-lang/src/main/scala/io/smartdatalake/meta/state/StateMigrator.scala (hunk 4 above)
Comment (zzeekk) [reply sdl-lang/src/main/scala/io/smartdatalake/meta/state/StateMigrator.scala:38]: Yes, sure. Done.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/StateMigratorDef.scala (hunk 3 above)
Comment (pgruetter) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/StateMigratorDef.scala]: OK, makes sense. Thanks!

**Review by pgruetter — APPROVED** [review]

**Comment by iliassoto** [comment]
I did some testing for this feature. The mentioned files are attached:

1. I ran "myApp.conf" and with SDLB 2.3.0., resulting in the state "myAppName.1.1.json"
2. I then created a new project with the SDLB built from this branch, and tried to run the pipeline again using the old state file, where an error was thrown (see file). 

The JSON itself is able to be parsed: 

```scala
  // read state from json
def fromJson(stateJson: String): ActionDAGRunState = {
      val jValue = JsonMethods.parse(stateJson) //WORKS
      val migratedJValue = checkStateFormatVersionAndMigrate(jValue).getOrElse(jValue) //ERROR HAPPENS HERE
      // extract into class structures
      migratedJValue.extract[ActionDAGRunState]
    } catch {
      case ex: Exception => throw new IllegalStateException(s"Unable to parse state from json: ${ex.getMessage}", ex)
    }
  }
```

[error.log](https://github.com/smart-data-lake/smart-data-lake/files/14001698/error.log)
[myApp.hocon.txt](https://github.com/smart-data-lake/smart-data-lake/files/14001699/myApp.hocon.txt)
[myAppName.1.1.json](https://github.com/smart-data-lake/smart-data-lake/files/14001700/myAppName.1.1.json)

The operation _json \ "runStateFormatVersion"_ results in a _JNothing_-Type, which throws the error during the assert.

```scala
  override def migrate(json: JValue): JValue = {
    assert(json \ "runStateFormatVersion" match { case JInt(version) => version <= versionFrom}, s"Version should be equals or less than $versionFrom")
    json.transformField {
      case (name, actionsState) if name == "actionsState" =>
        (name, actionsState.transformField {
          case (name, results) if name == "results" =>
            (name, JArray(results.children.map {
              result =>
```

I guess we can just add another case that also handles JNothing for versions that don't have that node in their JSON files?

**Comment by zzeekk** [comment]
> ```scala
> I guess we can just add another case that also handles JNothing for versions that don't have that node in their JSON files?
> ```
You're right, early state files without version attribute produce an error. I added the JNothing case, and fixed adding the version attribute afterwards.

**Review by iliassoto — APPROVED** [review]

**Pull request merged**